### ISSUE TITLE
Drop Emacs 25

### ISSUE THREAD

**Issue by jabranham** [emacs-ess/ESS#876]
How do people feel about announcing dropping support for Emacs 25 after the 19.04 release? Emacs 26.1 has been out about a year already, and Emacs 26.2 should be out any day now. A year is plenty of time to upgrade. And those who are stuck on 25 can continue to use 19.04. WDYT?

**Comment by lionel-** [comment]
I think 1 year is too short because people tend to delay their Emacs upgrades. What features do you need from 26.1?

**Comment by vspinu** [comment]
Way to early indeed. 27 is not released, nor even 26.2. Supporting 25 is not hard at all.

**Comment by mmaechler** [comment]
Yes!   Our users are *not* mainly emacs hackers and PhD students!
If you work in a bigger company (outside of IT), e.g., pharmaceutical (!),
you cannot easily update emacs at all.  Nor in a bank or insurance company, nor at FDA, or .. or ..

**Comment by jabranham** [comment]
@lionel- It'd be more like 1.5 years, since 19.04 would support 25 and the release after that (19.10?) wouldn't. But that's minor. I've already been bitten several times by Emacs 25 stuff, most recently with `cXXr` not existing in elisp until 26. It'd also be nice to be able to simplify the code in several places (especially the hacky way we deal with flymake right now since it requires 26). The optional arguments to `string-trim` would be nice to use too. And we could remove our backport of `ess-display-buffer-reuse-mode-window`. And that's just off the top of my head.

I still haven't heard a convincing case against dropping support for Emacs 25. @vspinu it's not hard, but it does create some uglier code than needed. And @mmaechler those users unable to update Emacs can live with the older ESS version.

But if we're keeping 25 support through at least the release after next, my question is when are we dropping it?

**Comment by vspinu** [comment]
> my question is when are we dropping it?

Shortly after emacs 28 is out. I think we should what most emacs projects do. Support last 2 stables of emacs.

**Comment by jabranham** [comment]
So let's say Emacs 27.1 is released May 2020 and ESS continues our current 6-month XX.04/10 release cycle.

Then ESS 20.10 mentions that next release we drop Emacs 25, then actually dropping support in 2021? Emacs 26 will be 3 years old at that point. I think it's a waste of our limited time to think about supporting people who insist on staying on old Emacses that long. If they want an old Emacs, they can run an old ESS.

**Comment by vspinu** [comment]
We can drop 25 in the same year when emacs 27 is released. So far I haven't seen projects dropping 25. Let's postpone this discussion for another year or two.

**Comment by gdkrmr** [comment]
I guess the main question is the following: Is there a need to do so? Do you need any functionality that is not available in emacs 25?

**Comment by lionel-** [comment]
Some data points:

- clojure-mode supports Emacs 25: https://github.com/clojure-emacs/clojure-mode/blob/master/.circleci/config.yml

- use-package supports Emacs 24.1: https://github.com/jwiegley/use-package/blob/master/.travis.yml

- evil-mode supports Emacs 24.5: https://github.com/emacs-evil/evil/blob/master/.travis.yml

- aggressive-indent supports Emacs 24.1: https://github.com/Malabarba/aggressive-indent-mode/blob/master/aggressive-indent.el#L8

- avy-mode supports Emacs 24.1: https://github.com/abo-abo/avy/blob/bbf1e7339eba06784dfe86643bb0fbddf5bb0342/avy.el#L8

- company-mode supports Emacs 24.5: https://github.com/company-mode/company-mode/blob/master/.travis.yml

- dash supports Emacs 24.1: https://github.com/magnars/dash.el/blob/master/.github/workflows/test.yml

- diff-hl supports Emacs 24.3: https://github.com/dgutov/diff-hl/blob/2281a89a3ddc6616073da6f190dda08d23b18ba6/diff-hl.el#L9

- elpy supports Emacs 24.3: https://github.com/jorgenschaefer/elpy/blob/master/.travis.yml

- flycheck supports Emacs 24.5: https://github.com/flycheck/flycheck/blob/master/.travis.yml

- js2-mode supports Emacs 24.1: https://github.com/mooz/js2-mode/blob/40aab27581279d0fdbfeb9afeb85f39d401a927f/js2-mode.el#L12

- magit supports Emacs 25.1: https://github.com/magit/magit/blob/master/.travis.yml

- markdown-mode supports Emacs 25.1: https://github.com/jrblevin/markdown-mode/blob/master/.github/workflows/test.yml

- php-mode supports Emacs 24.4: https://github.com/emacs-php/php-mode/blob/master/.github/workflows/test.yml

- projectile supports Emacs 25.1: https://github.com/bbatsov/projectile/blob/master/.travis.yml

- smartparens supports Emacs 25.1: https://github.com/Fuco1/smartparens/blob/master/.github/workflows/test.yml

Dropping Emacs 25 now would make us ahead of all these popular Emacs projects. Most of them still support Emacs 24.x.

I think we need a clear lifecycle duration, but I'm not sure what would be best. 5 years makes sense to me because that's how long we support old versions of R at RStudio for customers stuck on older enterprise environments. For reference, RHEL 7 released in 2014 ships Emacs 24.3. RHEL 8 released in 2019 ships Emacs 26.1.

**Comment by lionel-** [comment]
The last released version of ESS (18.10.2) still supports Emacs 24.3 (released in 2013). If we have a 20.09 release with support for Emacs 25.1 (released in 2016-09), this means we have a 4 years support.

**Comment by jabranham** [comment]
Is there a reason you want to support Emacs 25 still? It seems like all the major distros (save RHEL 7) ship with Emacs 26 so it's a safe bet that most people have at least 26.1.

I think 5 years is way too long. Rstudio is a company with paying customers. I develop ESS for fun in my spare time. It's not fun to support old versions of Emacs. If someone wants to pay me to do so, fine, otherwise I have little interest.

> If we have a 20.09 release

Which we probably won't, given there aren't any plans to do so (at least, not that I'm aware of)?

**Comment by lionel-** [comment]
> Is there a reason you want to support Emacs 25 still?

1. Dropping it now would make us ahead of most Emacs packages. I couldn't find an important package that has dropped 25.1 already.

2. The currently released version of ESS (the version likely used in controlled environments) supports a 7 years old Emacs. Shortening that to 2 years in one go seems risky.

Another relevant question is what do we gain by dropping Emacs 25.1 support? It doesn't seem like it's hard to support it but I may be missing something.

> I develop ESS for fun in my spare time.

I think we are all developing ESS in our spare time? This doesn't mean that the needs of ESS users have no priority.

> Which we probably won't, given there aren't any plans to do so (at least, not that I'm aware of)

Unless I'm mistaken all blocker issues have been solved so that could go quickly (ping @mmaechler). In any case we almost certainly shouldn't drop 25.1 support before that release.

**Comment by jabranham** [comment]
> Another relevant question is what do we gain by dropping Emacs 25.1 support?

I keep running into places where something works on Emacs 26+ but not 25: #989 (which solves an annoying bug reported by at least two users #1012 and #987) and `ess-display-buffer-reuse-mode-window` backport off the top of my head, probably more.

> I think we are all developing ESS in our spare time? 

I meant as opposed to Rstudio developers who get paid by customers who might want support for old versions of things.

> This doesn't mean that the needs of ESS users have no priority.

Which I never suggested? I'm still not convinced that anyone is using Emacs 25 - all the major distros, homebrew, etc have all updated to at least Emacs 26 aside from RHEL6 which is shipping Emacs 24 which we don't support anyway.

**Comment by lionel-** [comment]
> I keep running into places where something works on Emacs 26+ but not 25: #989 (which solves an annoying bug reported by at least two users #1012 and #987)

Thanks.

> ess-display-buffer-reuse-mode-window backport

There is probably no hurry to remove backports?

> I'm still not convinced that anyone is using Emacs 25

Surely this is a bit extreme. I have no idea about how many users are stuck on old Emacsen. This is why I'm relying on @mmaechler who knows our user base better, and on the little survey of supported versions of other popular Emacs packages I posted earlier.

> RHEL6 which is shipping Emacs 24 which we don't support anyway.

This is RHEL7, and we do support Emacs 24.3 right now. This is an important distinction, only the dev version requires Emacs 25.1. This is why dropping support before release would shorten our support from 7 years to 2 years in one go, which seems rather drastic.
